saml-idp.py

```diff
@@ -149,11 +149,43 @@ def make_assertion(resp, acs, irt, iss, username, validity):
                attrib={'InResponseTo': irt,
                        'NotOnOrAfter': tsnow(validity),
                        'Recipient': acs})
+
+    # minimal required for Okta to create a new user
+    attribute_statement = subelement(assertion, 'saml2:AttributeStatement')
+    email_attr = subelement(attribute_statement, 'saml2:Attribute', attrib={
+        "Name": "email", 
+        "NameFormat": "urn:oasis:names:tc:SAML:2.0:attrname-format:basic"
+        })
+    email_val = subelement(email_attr, "saml2:AttributeValue")
+    email_val.text=username
+
+    name_attr = subelement(attribute_statement, 'saml2:Attribute', attrib={
+        "Name": "firstName", 
+        "NameFormat": "urn:oasis:names:tc:SAML:2.0:attrname-format:basic"
+        })
+    name_val = subelement(name_attr, "saml2:AttributeValue")
+    name_val.text="First"
+
+    last_attr = subelement(attribute_statement, 'saml2:Attribute', attrib={
+        "Name": "lastName", 
+        "NameFormat": "urn:oasis:names:tc:SAML:2.0:attrname-format:basic"
+        })
+    last_val = subelement(last_attr, "saml2:AttributeValue")
+    last_val.text="Last"
+    
+    # optional uid field, which can get saved as external_user_id if mapped to do so in Okta
+    uid_attr = subelement(attribute_statement, 'saml2:Attribute', attrib={
+        "Name": "uid", 
+        "NameFormat": "urn:oasis:names:tc:SAML:2.0:attrname-format:basic"
+        })
+    uid_val = subelement(uid_attr, "saml2:AttributeValue")
+    uid_val.text=uid
+
     cond = subelement(assertion, 'saml2:Conditions',
                       attrib={'NotBefore': tsnow(-validity),
                               'NotOnOrAfter': tsnow(validity)})
     aud_rest = subelement(cond, 'saml2:AudienceRestriction')
-    subelement(aud_rest, 'saml2:Audience').text = acs  # todo urlparse
+    subelement(aud_rest, 'saml2:Audience').text = sp_entity_id
```

Comment:
I don't know what `acs` is but should this replace `acs` or should it be added along with it?

Reply:
I lot of this is a bit mirky for me too... but seems like `acs` here is the url that okta tells the idp to send the client back to with the token, but the Audience is like a unique identifier of the integration on the service provider (okta).  It looks like in some cases they are the same value so thats why the author of this tool used the same variable..  But with okta they are different so i had to make this change..  Which took a few hours to figure out 😬 .  You'll see on line 151 the other place that `acs` is still used.